services/app-api/src/resolvers/contract/withdrawAndReplaceRedundantRate.ts

```diff
@@ -0,0 +1,82 @@
+import { ForbiddenError } from 'apollo-server-lambda'
+import { isAdminUser } from '../../domain-models'
+import type { MutationResolvers } from '../../gen/gqlServer'
+import { logError, logSuccess } from '../../logger'
+import type { Store } from '../../postgres'
+
+import {
+    setErrorAttributesOnActiveSpan,
+    setResolverDetailsOnActiveSpan,
+    setSuccessAttributesOnActiveSpan,
+} from '../attributeHelper'
+import { GraphQLError } from 'graphql'
+
+/**
+ * withdrawAndReplaceRedundantRate
+ * @param store
+ * @returns an updated contract
+ *
+ * withdrawAndReplaceRedundantRate
+ * this is an admin only action
+ * replaces an existing child rate with a linked rate
+ * and in the process, marks old child rate as withdrawn and logs activity in changelog for both contract and rates
+ */
+
+export function withdrawAndReplaceRedundantRateResolver(
+    store: Store
+): MutationResolvers['withdrawAndReplaceRedundantRate'] {
+    return async (_parent, { input }, context) => {
+        const { user, ctx, tracer } = context
+        const span = tracer?.startSpan(
+            'withdrawAndReplaceRedundantRate',
+            {},
+            ctx
+        )
+        setResolverDetailsOnActiveSpan(
+            'withdrawAndReplaceRedundantRate',
+            user,
+            span
+        )
+
+        const {
+            contractID,
+            replaceReason,
+            withdrawnRateID,
+            replacementRateID,
+        } = input
+        span?.setAttribute('mcreview.package_id', contractID)
+
+        if (!isAdminUser(user)) {
+            const message =
+                'user not authorized to withdraw or replace rate on contract'
+            logError('withdrawAndReplaceRedundantRate', message)
+            setErrorAttributesOnActiveSpan(message, span)
+            throw new ForbiddenError(message)
+        }
+
+        const withdrawAndReplaceRedundantRateResult =
+            await store.replaceRateOnContract({
+                contractID: contractID,
+                replacementRateID,
+                withdrawnRateID,
+                replaceReason,
+                replacedByUserID: user.id,
+            })
+        if (withdrawAndReplaceRedundantRateResult instanceof Error) {
+            const errMessage = `Failed to unlock contract revision with ID: ${contractID}; ${withdrawAndReplaceRedundantRateResult.message}`
+            logError('withdrawAndReplaceRedundantRate', errMessage)
+            setErrorAttributesOnActiveSpan(errMessage, span)
+            throw new GraphQLError(errMessage, {
```

Comment:
There's a few errors we can get here that I think are UserInputErrors, but we're returning everything as an INTERNAL_SERVER_ERROR here. 

I think it's worth making a new error type that we can differentiate on. We already have a NotFoundError in postgresErrors, we could have something else for some of these other user input errors.

Reply:
@macrael Yes I was thinking that too but forgot about it - thanks for calling this out! I do want the errors to be more distinct coming out of the API - will look at this today